### DISCUSSION TITLE
Custom executor - uniqueness of TaskInstanceKey and sensor tasks in reschedule mode

### DISCUSSION THREAD

**Discussion by michalc** [apache/airflow#35229] [Q&A]
I'm writing a custom executor, and I made a bit of a mis-assumption about the uniqueness of TaskInstanceKey. I thought that execute_async will be called just the once for each TaskInstanceKey (assuming it doesn't raise an exception or anything), and so TaskInstanceKey could be stored as metadata on the remote system that executes tasks, and then I can use this metadata in the sync and in try_adopt_task_instances functions.

However, I think that sensor tasks violate this assumption(?) - when rescheduled, the next time they run execute_async will be called again for the same TaskInstanceKey. Which then means the sync function and try_adopt_task_instances functions can't really identify the right "run" in the remote system because there could be multiple that match the TaskInstanceKey. And then I think from the sync(?) function errors like the following happen, because it's reporting on the wrong thing:

> Executor reports task instance %s finished (%s) although the task says it's %s. (Info: %s) Was the task killed externally?

So what to do? Is there some sort of metadata available in execute_async that can be sent to the remote system to then properly identify the right runs in the sync and try_adopt_task_instances functions?

**Comment by potiuk** [comment]
cc: @o-nikolas - maybe you can have some insights, it must have been discussed as part of AWS executor implementation (or maybe that's the case you YET need to handle :P :) ?

**Reply by michalc** [reply]
Now am doing this:

- keeping the unique ID in the remote system of the execution as state on the executor, and using that in the sync function to make sure the right remote state is used in change_state
- to get the correct remote ID in the case of try_adopt_task_instances, to choose between multiple ones that all match a given TaskInstanceKey, choosing the one that has the most recent updated at time stored in the remote system

But I guess feels a bit odd somehow - should there be a unique and persistent ID for each "trigger of execute_async" to link the two systems together more easily?
